arm-software/shared/elf2bin/elf2bin.cpp

```diff
@@ -0,0 +1,468 @@
+    }
+    segments_wanted = bases;
+  }
+
+  uint64_t bankwidth = 1, nbanks = 1;
+  if (Arg *A = Args.getLastArg(OPT_banks_EQ)) {
+    StringRef str = A->getValue();
+    size_t xpos = str.find_first_of("x");
+    size_t xpos2 = str.find_last_of("x");
+    if (!(xpos == xpos2 && !str.substr(0, xpos).getAsInteger(0, bankwidth) &&
+          !str.substr(xpos + 1).getAsInteger(0, nbanks) &&
+          llvm::has_single_bit(bankwidth) && llvm::has_single_bit(nbanks) &&
+          bankwidth * nbanks != 0))
+      fatal(Twine("cannot parse bank specification '") + str + "'");
+  }
+
+  uint64_t datareclen = 16;
+  if (Arg *A = Args.getLastArg(OPT_datareclen_EQ)) {
+    StringRef str = A->getValue();
+    if (str.getAsInteger(0, datareclen))
+      fatal(Twine("cannot parse data record length '") + str + "'");
+  }
+
+  bool include_zi = Args.hasArg(OPT_zi);
+  bool physical = true;
+  if (Arg *A = Args.getLastArg(OPT_physical, OPT_virtual))
+    physical = A->getOption().matches(OPT_physical);
+
+  if (infiles.empty())
+    fatal_suggest_help("no input file specified");
+  if (!outfile && !outpattern)
+    fatal_suggest_help("no output filename or pattern specified");
+  if (!format)
+    fatal_suggest_help("no output format specified");
+  if (outfile && outpattern)
+    fatal_suggest_help("output filename and pattern both specified");
+
+  if ((format != Format::BinCombined && format != Format::VhxCombined) &&
+      baseaddr)
+    fatal("--base only applies to --bincombined and --vhxcombined");
+
+  if ((format != Format::BinMultifile && format != Format::BinCombined &&
+       format != Format::VhxMultifile && format != Format::VhxCombined) &&
+      (bankwidth != 1 || nbanks != 1))
+    fatal("--banks only applies to binary and VHX output");
+
+  /*
+   * Open the input files.
+   */
+  std::vector<InputObject> objects;
+
+  for (const std::string &infile : infiles) {
+    objects.emplace_back();
+    InputObject &inobj = objects.back();
+    inobj.filename = infile;
+
+    ErrorOr<std::unique_ptr<MemoryBuffer>> membuf_or_err =
+        MemoryBuffer::getFileOrSTDIN(infile, false, false);
+    if (std::error_code error = membuf_or_err.getError())
+      fatal(infile, "unable to open", errorCodeToError(error));
+    inobj.membuf = std::move(membuf_or_err.get());
+
+    Expected<std::unique_ptr<llvm::object::Binary>> binary_or_err =
+        llvm::object::createBinary(inobj.membuf->getMemBufferRef(), nullptr,
+                                   false);
+    if (!binary_or_err)
+      fatal(infile, "unable to process", binary_or_err.takeError());
+
+    std::unique_ptr<ELFObjectFileBase> elf =
+        dyn_cast<ELFObjectFileBase>(*binary_or_err);
+    if (!elf)
+      fatal(infile, "unable to process: not an ELF file");
+    inobj.elf = std::move(elf);
+  }
+
+  /*
+   * Helper function for listing the segments of a file, paying
+   * attention to the --segments option to restrict to a subset.
+   */
+  auto segments = [&](InputObject &inobj) {
```

Comment:
Do we check that the ELF file isn't relocatable and error out with something like "ELF file must be type ET_EXEC or ET_DYN"?

Reply:
Not explicitly, but we check there's a program header table, which surely has the same effect.

Reply:
OK, I see that check in a different file. I'll put a comment on that line to see if we can make the error message less ELF specific. Just thinking that a user might not know that relocatable objects don't have program headers. I realise that they are even less likely to understand ET_EXEC and ET_DYN.